### ISSUE TITLE
Mushroom Generator immediate needs

### ISSUE THREAD

**Issue by elanhickler** [RobinSchmidt/RS-MET#118]
in order of priority

1. ~On project load, preset does not load correctly: http://www.elanhickler.com/transfer/MushroomGenerator.rpp (move Master Rate slider, and suddenly the patch works again.)~
1. ~Save grid settings for breakpoint envelope generators.~
1. ~Meta parameter renaming based on connection: https://github.com/RobinSchmidt/RS-MET/issues/110~
1. GUI needs update on preset change: https://github.com/RobinSchmidt/RS-MET/issues/113
1. ~`presetPath` path is not being opened when clicking "load" button for preset widget, or at least it wasn't working last time I checked: https://github.com/RobinSchmidt/RS-MET/issues/39~
1. ~Smoothing + Modulation does not work properly on child modules (LFOs/ADSRs). When setting a slider with modulation, the modulation ceases while smoothing. I will disable smoothing on child modules for now.~

I am not waiting for these fixes to release Mushroom Generator, but user complaints will start coming in if these are not dealt in the next week. Would be greatly appreciated if you dealt with these in the next few days.

**Comment by RobinSchmidt** [comment]
yes, i will look into this tomorrow

**Comment by elanhickler** [comment]
btw about smoothing speed, it's not about the speed per slider, actually the problem is that exponential variables smooth fast going up and slow going down. This is not noticeable when smoothing speed is low, and this is the last problem I care about solving at the moment, but just want to let you know if you are thinking of improving the smoother. You have to smooth the slider's "position" rather than applying smoothing on the actual value being outputted.

**Comment by elanhickler** [comment]
Oh, also smoothing speed is only set if the smoother is first finished smoothing. Would it be an easy fix to set smoothing time to currently-smoothing parameters?

**Comment by elanhickler** [comment]
I just implemented by own function for smoothing setting smoothing:

![image](https://user-images.githubusercontent.com/9116575/32561565-c2aad4ce-c469-11e7-8a02-3783de56c231.png)

but the smoother jumps to target value when calling setTimeConstantAndSampleRate, whyyyyy

**Comment by elanhickler** [comment]
new bug, see #1.

**Comment by RobinSchmidt** [comment]
> I just implemented by own function for setting smoothing

what is that function supposed to do? i ask because probably it doesn't do what you think it does. the 
 usedSmoothers array are only the currently active/running smoothers. how can you access the array anyway? it's a protected variable.

**Comment by elanhickler** [comment]
yeah I had to make it not protected 😄 👍 

The active smoothers need their smoothing times set as well, when setting smoothing time.

**Comment by RobinSchmidt** [comment]
>  I had to make it not protected. The active smoothers need their smoothing times set as well, when setting smoothing time.

my point is that you seem to be hacking into my code without really knowing what you are doing and then at some point you will wonder why things don't work as you expect. seems like your goal is to set the smoothing time for all smoothers globally? then just iterating over the usedSmoothers array won't achieve that. even if you iterate the same way over smootherPool too won't be enough because if there aren't enough smoothers in the pool a new one will be created - and that will then not have the smoothing time, you think it should have

**Comment by elanhickler** [comment]
After removing my "hack" preset loading still doesn't work.

....also you didn't see my entire function... and if you did see my entire function you would know I am doing exactly what needs to be done to achieve my goal:

![image](https://user-images.githubusercontent.com/9116575/32572533-1fc0a65c-c488-11e7-8dd4-7bc403cb45e5.png)

**Comment by RobinSchmidt** [comment]
> After removing my "hack" preset loading still doesn't work

yes, i wasn't saying that your hack is the reason for the bug you reported. the problem is that the contents of the usedSmoothers and smootherPool arrays is not fixed. new ones may be added dynamically as desired. and these new ones will then not respect your intended global setting. if you want to set global smoothing speed, better iterate over the parameters and call their setSmoothingTime function

.

**Comment by elanhickler** [comment]
the new ones do respect my global setting because I have already realized that the smoothing time is derived from the parameter, not the smoother pool.

and yes, that's what my image depicts, I call setSmoothingTime

**Comment by RobinSchmidt** [comment]
no, you call rsSmoother::setTimeConstantAndSampleRate - you should instead call rsSmoothingTarget::setSmoothingTime on the parameter (its a subclass of rsSmoothingTarget)

**Comment by elanhickler** [comment]
I am doing that already!

I am doing both!

Robin, I need to get Mushroom Generator released, so yes I will hack your code if it means I don't have to wait another day for you to fix something.

**Comment by elanhickler** [comment]
Smoothing works. ... wait... wtf...

maybe preset loading only doesn't work on parameters with modulation.... but then it only doesn't work on master rate... ughhhhhhhhhhh, there might be a bigger issue here... unless you can figure it out easily.

**Comment by RobinSchmidt** [comment]
> I am doing that already! I am doing both!

ah. hmm..ok, i see. i suppose, the iteration over the usedSmoothers is then just to let the new setting take effect immediately for those smoothers that are currently already running?

**Comment by elanhickler** [comment]
yesssssssssssssss

**Comment by RobinSchmidt** [comment]
ok, yes, that should work. but from that snippet alone, it seemed to me that something is wrong there

**Comment by elanhickler** [comment]
http://www.elanhickler.com/transfer/PLUCK_OrganicBounce.xml

I'm trying to figure out a pattern as to why this preset doesn't load. Everything seems to be working in except the master rate on this particular patch. Going to keep trying to reproduce the bug.

**Comment by elanhickler** [comment]
Ok, I found one consistency: If you move the Master Rate any one time, then the OrganicBounc preset loads correctly, and everything seems to work moving forward.

**Comment by RobinSchmidt** [comment]
hmm...have you somewhere overriden the setStateFromXml function and maybe setting this particular parameter there? ...just a quick guess off the cuff

**Comment by elanhickler** [comment]
Running the debugger, when I do ModulatableParameter2::getValue() it returns the correct number 1.

I'm triyng to do a temporary fix to set the FreqMult parameter ("Master Rate"), but setting the value seems not to have an effect.

.. WAIT!! I think i found it.

![image](https://user-images.githubusercontent.com/9116575/32574257-dd9228d6-c48d-11e7-80ac-99fe24e90910.png)


Disabling that if statement helped. Maybe the if statement needs to be moved to just setUnmodulatedValue, as setValue already has that clause.

**Comment by RobinSchmidt** [comment]
> Disabling that if statement helped. 

aahh. ok...hmm...yes...thinking about it, that if statement may be wrong. with it there may be no consistency between the Parameter::value and ModulatableParameter::unmodulatedValue. ok, i'll remove that "if".

i think that modulatedValue = unmodulatedValue might not be strictly necessarry no anymore. but maybe it's a good idea anyway to keep them consistent in case of no modulations connected

**Comment by RobinSchmidt** [comment]
> btw about smoothing speed, it's not about the speed per slider, actually the problem is that exponential variables smooth fast going up and slow going down.

hmm...ok, i see. that's indeed the problem of smoothing the final, "mapped" value instead of the normalized 0..1 value

**Comment by elanhickler** [comment]
![image](https://user-images.githubusercontent.com/9116575/32586941-ecc22326-c4c2-11e7-8c12-e7d161748b52.png)

now this is causing a problem.

Somehow you are setting value without calling the callbacks, so now callbacks at construction are not called causing certain problems.

For example, my modulation sources never have a chance to set the display name:

![image](https://user-images.githubusercontent.com/9116575/32586984-35778890-c4c3-11e7-90d6-9fad8af1c64d.png)

I solved the problem for my LFOs, but haven't figured it out for the midi sources yet, weird...

**Comment by elanhickler** [comment]
Ok, I found the problem. You should force callbacks to be called when setting default values on construction. Or somehow make sure value == newValue is false.

Debugger video: https://puu.sh/yhWNc/bcb23dae42.mp4

**Comment by RobinSchmidt** [comment]
just working on storing/recalling the snap and grid settings. this turns out to be more complicated than i thought...

**Comment by elanhickler** [comment]
I kinda figured it would be complicated actually... Do whatever or don't do it, I'll be fine for now.

**Comment by RobinSchmidt** [comment]
i'll do it. i'll write one work-hour for it, whatever it takes. it took me more time already, but it should go mostly on my account. i want it too

**Comment by RobinSchmidt** [comment]
it seems to work now. i have just to make it work for the editor for multiple breakpoint-modulators, too (the one i use in straightliner that shows multiple curves in one plot). but for the regular, single-curve editor, it seems to work

**Comment by elanhickler** [comment]
Great!

I just got a user feature requests for renaming the "Meta ##" based on attachment.

Told you!

Sales are going well, so I guess you should be charging me for time spent on my company's needs. I may need you to look into some bugs with interface. I'll make a new issue for these specific things soon. Keep working on my "immediate needs" list if possible!

**Comment by RobinSchmidt** [comment]
yes, i'll try to sort out the remaining issues on the multi-envelope editor and when done work off the bullet points on the list

**Comment by RobinSchmidt** [comment]
ok, i think, i have fixed it - will next do the meta-renaming stuff

**Comment by RobinSchmidt** [comment]
hmm...i have implemented the meta name change now, but (at least in the juce) host, the new names are shown in the host only after reloading the project. it seems, calling `updateHostDisplay()` is not enough. the juce host does not try to inquire a new name when i call this method - and there seems to be no other method than that

**Comment by elanhickler** [comment]
Can I keep adding things to the list? I'm going to add something for presets: make sure preset path is working. On mac, I don't think my plugins open the specified `presetPath` path when clicking the "load" button.

**Comment by RobinSchmidt** [comment]
ok, keep adding. today i fixed a lot of problems in FuncShaper. seems to work well now. i'm leaving now for until evening. i think, i'll try to look into gui update thing tomorrow

**Comment by elanhickler** [comment]
robin there is actually still a preset switching bug in mushroom generator. I can't get the details for you at the moment.

**Comment by elanhickler** [comment]
here's a user report:

### Chris "Jerobeam" Fenderson
> - When skipping through the presets back and forth, it sometimes seems to forget to change some parameters. Example: FX_HarmonicLocking, parameters Density and Tune are different, depending if I'm switching forward from FX_GasPoweredSpirals or back from FX_HarmonicLocking2. Also sometimes the name ends with a * and sometimes it doesn't, but that doesn't always seem to correlate with forgotten parameters.

**Comment by elanhickler** [comment]
robin, please look at the preset path, im sure it's an easy fix.

**Comment by RobinSchmidt** [comment]
ok, i just need to finish my iir filter migration then i'll look into it

**Comment by RobinSchmidt** [comment]
i tried to look into the preset path thing but i can't build mushroom generator anymore because i rename a few classes in my library (just added the rs prefixes). can you update?

**Comment by elanhickler** [comment]
oh yeah, you won't be able to build shroom even after that, why wouldn't you just use your mac chainer?....... oh are you not using proper preset paths? tsk tsk.

**Comment by elanhickler** [comment]
edit: actually for now you should just change your preset path on mac chainer to anything other than the plugin's folder... or even try it on windows chainer. This problem is universal.

**Comment by RobinSchmidt** [comment]
> even try it on windows chainer. This problem is universal.

oh - then i probably don't understand what the problem is. i don't have any problems with the preset path in my chainer on windows

**Comment by RobinSchmidt** [comment]
i'm using a path "Presets" that sits directly below the folder with the plugin dll. this then contains subfolders "Starightliner", "FuncShaper", etc. - there are all my presets. it works without problems here - assuming the path exists, of course

**Comment by elanhickler** [comment]
ok, then try a different path, any path that is not next to the plugin.

**Comment by elanhickler** [comment]
ok you can build my project if you make smoothingManager->usedSmoothers a public member.

**Comment by RobinSchmidt** [comment]
i had to add the rs-prefix to the elliptic filter declaration in your code. did you not update my library or how could you build it? now, when i copy the "Presets" folder into the directory where the dll is and click on load, it opens the correct folder. where is the problem?

**Comment by elanhickler** [comment]
robin PLEASE TEST.........omgdf.gdfgusodfnjg!!

the problem is that I do not want the preset folder next to the dll.

You must change the preset folder location and make sure the specified location is working. ANY location other than the preset folder next to the dll.

**Comment by elanhickler** [comment]
as for renaming rs prefix, I must be having some trouble with git, trying to do something fancy so I can a pull request for only specific things, like the metaparameter renaming edit.

**Comment by RobinSchmidt** [comment]
so, you mean, i should edit the strings that specify the preset folder, then move the preset folder to that location and then try again and then it will stop to work? what do you suggest i should use a preset folder location on windows?

**Comment by elanhickler** [comment]
OR use mac which already has a specified different folder but the folder not existing may cause an issue 

try destkop/presets on windows or somsethings

**Comment by RobinSchmidt** [comment]
ok, i'm setting it now in the mushroom constructor to the user's documents directory. still works

**Comment by elanhickler** [comment]
it has never worked on mac............

**Comment by elanhickler** [comment]
ok i need to test on windows myself

**Comment by elanhickler** [comment]
what if 

`#elif __APPLE__`

is not working? the apple macro

**Comment by elanhickler** [comment]
yep works for me on windows, i bet it's the macro.

**Comment by RobinSchmidt** [comment]
```
#elif __APPLE__
	presetPath     = File("~/Library/Audio/Presets/"+moduleVendor+"/"+moduleName+"/Presets").getFullPathName();
```
...is this actually correct? ..i mean with the tilde and stuff. and you need to fill in this +moduleVendor+ placeholders as well, me thinks

**Comment by RobinSchmidt** [comment]
can you look at the presetPath string in the debugger and compare with the actual path on your mac?

**Comment by elanhickler** [comment]
the problem was the ~

I forgot, was confused back when I set this up. ~ Should only be for user presets, not factory.

**Comment by elanhickler** [comment]
Ok, 2 more major bugs. 4 and 6.

**Comment by RobinSchmidt** [comment]
i just want to integrate into my filter design procedures the halpern and proper gaussian approximation methods (i need to do some refactoring before that). then i have a nice bunch of IIR design methods to choose from to make filter-banks (butterworth, linkwitz-riley, bessel, papoulis, chebychev 1/2, elliptic, halpern and gauss)

**Comment by RobinSchmidt** [comment]
> Smoothing + Modulation does not work properly on child modules

ok, i just changed quite a lot about the mod-system to make it work at all in quadrifex (it currently works only in the waveshaper effect there - you can try it with the DC parameter). since the waveshaper or any module you insert in quadrifex becomes a child-module of quadrifex (which it itself a child-module of the chainer, in this case)....and it seems to work there now, it may now work for mushroom, too (i hope)

**Comment by elanhickler** [comment]
Works on spiral generator! 

One more major issue!
